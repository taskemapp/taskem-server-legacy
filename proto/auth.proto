syntax = "proto3";

package auth;

service Auth {
    rpc SignUp (SignupRequest) returns (SignUpResponse);
    rpc Login (LoginRequest) returns (LoginResponse);
    rpc Logout (LogoutRequest) returns (LogoutResponse);
}

message LoginRequest {
    string email = 1;
    string password = 2;
}

message LoginResponse {
    bool success = 1;
    string message = 2;
    string session_id = 3;
}

message LogoutRequest {
    string session_id = 1;
}

message LogoutResponse {
    string message = 2;
}

message SignupRequest {
    string email = 1;
    string user_name = 2;
    string password = 3;
}

message SignUpResponse {
    string message = 2;
}