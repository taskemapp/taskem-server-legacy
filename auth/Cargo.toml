[package]
name = "auth"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
argon2 = "0.6.0-pre.0"
db = { path = "../libs/db" }
diesel = { version = "2.1.6", features = ["postgres", "r2d2"] }
autometrics = { version = "1.0.1", features = ["prometheus-exporter"] }
tokio = { version = "1.0.0", features = ["rt", "rt-multi-thread", "macros"] }
prost = "0.12.4"
prost-types = "0.12.3"
tonic = { version = "0.11.0", features = ["tls", "default"] }
dotenv = "0.15.0"
axum = "0.7.5"
axum-server = "0.6.0"
tracing-subscriber = "0.3.18"
tracing = "0.1.40"
tower-service = "0.3.2"
tower = "0.4.13"
hyper = "1.3.1"
derive-new = "0.6.0"
r2d2_redis = "0.14.0"
regex = "1.10.4"
uuid = { version = "1.8.0", features = ["v4", "v7"] }
chrono = "0.4.38"

[build-dependencies]
tonic-build = "0.11.0"