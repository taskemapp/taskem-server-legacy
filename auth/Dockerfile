FROM rust:1.76.0 as build

WORKDIR /app

RUN apt update && apt install -y protobuf-compiler libprotobuf-dev libpq5 libpq-dev clang cmake gcc g++

COPY . .

RUN cargo build --release

FROM gcr.io/distroless/cc-debian12

# Set the architecture argument (arm64, i.e. aarch64 as default)
# For amd64, i.e. x86_64, you can append a flag when invoking the build `... --build-arg "ARCH=x86_64"`
ARG ARCH=aarch64

# libpq related (required by diesel)
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libpq.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libgssapi_krb5.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libldap-2.5.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libkrb5.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libk5crypto.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libkrb5support.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/liblber-2.5.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libsasl2.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libgnutls.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libp11-kit.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libidn2.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libunistring.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libtasn1.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libnettle.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libhogweed.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libgmp.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /usr/lib/${ARCH}-linux-gnu/libffi.so* /usr/lib/${ARCH}-linux-gnu/
COPY --from=build /lib/${ARCH}-linux-gnu/libcom_err.so* /lib/${ARCH}-linux-gnu/
COPY --from=build /lib/${ARCH}-linux-gnu/libkeyutils.so* /lib/${ARCH}-linux-gnu/

COPY --from=build /app/.env /.env
COPY --from=build /app/migrations /migrations
COPY --from=build /app/target/release/taskem /taskem

CMD ["/taskem"]
