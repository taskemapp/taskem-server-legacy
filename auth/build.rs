fn main() -> Result<(), Box<dyn std::error::Error>> {
    tonic_build::configure()
        .build_server(true)
        .protoc_arg("-I=../proto")
        .compile(&["auth.proto"], &["proto"])?;

    Ok(())
}
