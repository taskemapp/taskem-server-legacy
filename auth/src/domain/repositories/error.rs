#[derive(Clone, Debug)]
pub struct RepositoryError {
    pub message: String,
}

#[derive(Clone, Debug)]
pub struct SessionRepositoryError {
    pub message: String,
}
