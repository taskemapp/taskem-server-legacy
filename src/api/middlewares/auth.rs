use crate::domain::constants::MIDDLEWARE_AUTH_SESSION_KEY;
use crate::domain::repositories::session::RedisSessionRepository;
use derive_new::new;
use hyper::StatusCode;
use std::pin::Pin;
use std::sync::Arc;
use std::task::{Context, Poll};
use tonic::body::BoxBody;
use tonic::codegen::http::{Request, Response};
use tonic::transport::Body;
use tower::Layer;
use tower_service::Service;
use tracing::{debug, info};

#[derive(Clone, new)]
pub struct AuthMiddlewareLayer {
    redis_repository: Arc<dyn RedisSessionRepository>,
}

impl<S> Layer<S> for AuthMiddlewareLayer {
    type Service = AuthMiddleware<S>;

    fn layer(&self, inner: S) -> Self::Service {
        AuthMiddleware {
            inner,
            redis_repository: self.redis_repository.clone(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct AuthMiddleware<S> {
    inner: S,
    redis_repository: Arc<dyn RedisSessionRepository>,
}

type BoxFuture<'a, T> = Pin<Box<dyn std::future::Future<Output = T> + Send + 'a>>;

impl<S> Service<Request<Body>> for AuthMiddleware<S>
where
    S: Service<Request<Body>, Response = Response<BoxBody>> + Clone + Send + 'static,
    S::Future: Send + 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx).map_err(Into::into)
    }

    fn call(&mut self, req: Request<Body>) -> Self::Future {
        let clone = self.inner.clone();
        let mut inner = std::mem::replace(&mut self.inner, clone);

        debug!("Check auth url for request: \n{:?}", req.uri());

        if req.uri().path().contains("Auth") {
            return Box::pin(async move {
                let response = inner.call(req).await?;
                Ok(response)
            });
        }

        let headers = req.headers();

        debug!("Find authorization header");

        if let Some(header_value) = headers.get(MIDDLEWARE_AUTH_SESSION_KEY) {
            let value = header_value
                .to_str()
                .expect("Can't convert header value to str");
            let session = format!("session_id:{}", value);

            if self.redis_repository.session_expand(&session).is_ok() {
                info!("Session expanded");
                return Box::pin(async move {
                    let response = inner.call(req).await?;
                    Ok(response)
                });
            }
        }

        debug!("Unauthorized");

        Box::pin(async move {
            let res = Response::builder()
                .status(StatusCode::UNAUTHORIZED.as_u16())
                .body(BoxBody::default())
                .unwrap();
            Ok(res)
        })
    }
}

impl<S> Service<axum::extract::Request> for AuthMiddleware<S>
where
    S: Service<axum::extract::Request, Response = axum::response::Response> + Send + 'static,
    S::Future: Send + 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, request: axum::extract::Request) -> Self::Future {
        let headers = request.headers();

        debug!("Find authorization header");

        if let Some(header_value) = headers.get(MIDDLEWARE_AUTH_SESSION_KEY) {
            let value = header_value
                .to_str()
                .expect("Can't convert header value to str");
            let session = format!("session_id:{}", value);

            if self.redis_repository.session_expand(&session).is_ok() {
                info!("Session expanded");
                let future = self.inner.call(request);
                return Box::pin(async move {
                    let response: axum::response::Response = future.await?;
                    Ok(response)
                });
            }
        }

        debug!("Unauthorized");
        Box::pin(async move {
            Ok(axum::http::Response::builder()
                .status(StatusCode::UNAUTHORIZED)
                .body(axum::body::Body::empty())
                .unwrap())
        })
    }
}
