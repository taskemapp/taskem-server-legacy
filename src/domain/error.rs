use std::fmt::Display;

use derive_more::From;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, From)]
pub enum Error {
    RepositoryError,
    RedisError,
    FileError,
    GetPoolError,
    MapFromError,
}

impl Display for Error {
    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(fmt, "{self:?}")
    }
}
