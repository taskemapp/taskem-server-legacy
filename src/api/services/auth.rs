use std::sync::Arc;

use autometrics::autometrics;
use derive_new::new;
use tonic::{Request, Response, Status};
use tracing::log::error;

use crate::auth::auth_server::Auth;
use crate::auth::{
    LoginRequest, LoginResponse, LogoutRequest, LogoutResponse, SignUpResponse, SignupRequest,
};
use crate::core::regex::CachedRegexValidator;
use crate::domain::models::user::login_information::LoginInformation;
use crate::domain::models::user::user_information::UserInformation;
use crate::domain::repositories::session::RedisSessionRepository;
use crate::domain::repositories::user::UserRepository;

#[derive(new)]
pub struct AuthServiceImpl {
    pub(self) regex_cache: Arc<CachedRegexValidator>,
    pub(self) user_repository: Arc<dyn UserRepository>,
    pub(self) redis_repository: Arc<dyn RedisSessionRepository>,
}

#[autometrics]
#[tonic::async_trait]
impl Auth for AuthServiceImpl {
    async fn sign_up(
        &self,
        request: Request<SignupRequest>,
    ) -> Result<Response<SignUpResponse>, Status> {
        let sign_up_request = request.into_inner();

        if self
            .regex_cache
            .check_email(&sign_up_request.email)
            .is_err()
        {
            return Err(Status::invalid_argument("Invalid email"));
        }

        let user_repository = self.user_repository.clone();

        user_repository
            .create(&UserInformation::from(sign_up_request.clone()))
            .map_err(|e| {
                error!("{:?}", e);
                Status::internal("User creation failed")
            })?;

        Ok(Response::new(SignUpResponse {
            message: String::from("User successfully created"),
        }))
    }

    async fn login(
        &self,
        request: Request<LoginRequest>,
    ) -> Result<Response<LoginResponse>, Status> {
        let login_request = request.into_inner();

        if self.regex_cache.check_email(&login_request.email).is_err() {
            return Err(Status::invalid_argument("Invalid email"));
        }

        let user_repository = self.user_repository.clone();

        match user_repository.login(&LoginInformation::from(login_request)) {
            Ok(value) => {
                let session = self.redis_repository.create(&value.id).unwrap();
                return Ok(Response::new(LoginResponse {
                    user_name: value.user_name,
                    message: "User successfully logged".to_string(),
                    session_id: session,
                }));
            }
            Err(e) => {
                error!("{:?}", e);
                Err(Status::not_found("User not found"))
            }
        }
    }

    async fn logout(
        &self,
        request: Request<LogoutRequest>,
    ) -> Result<Response<LogoutResponse>, Status> {
        let logout_request = request.into_inner();
        match self
            .redis_repository
            .remove_session(&logout_request.session_id)
        {
            Ok(_) => Ok(Response::new(LogoutResponse {
                message: "User session removed".to_string(),
            })),
            Err(e) => Err(Status::internal(format!("Internal Server Error: {}", e))),
        }
    }
}
