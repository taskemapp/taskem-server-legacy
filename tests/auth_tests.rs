mod common;

#[cfg(test)]
mod auth_tests {
    mod auth {
        tonic::include_proto!("auth");
    }

    use std::error::Error;

    use auth::*;
    use insta::{assert_debug_snapshot, assert_yaml_snapshot};

    #[tokio::test]
    async fn sign_up() -> Result<(), Box<dyn Error>> {
        let channel = tonic::transport::Channel::from_static("http://0.0.0.0:50051")
            .connect()
            .await
            .unwrap();
        let mut client = auth_client::AuthClient::new(channel);

        let sign_up_req = SignupRequest {
            email: String::from("test@test.com"),
            user_name: String::from("test"),
            password: String::from("test"),
        };

        let response = client.sign_up(sign_up_req).await.unwrap();
        let inner = response.into_inner();
        assert_eq!(inner.message, "User successfully created");
        assert_debug_snapshot!(inner);

        Ok(())
    }

    #[tokio::test]
    async fn login() -> Result<(), Box<dyn Error>> {
        let channel = tonic::transport::Channel::from_static("http://0.0.0.0:50051")
            .connect()
            .await
            .unwrap();
        let mut client = auth_client::AuthClient::new(channel);

        let email = "test-login@test.com";
        let pass = "test";
        let user_name = "test-login";

        let sign_up_req = SignupRequest {
            email: String::from(email),
            user_name: String::from(user_name),
            password: String::from(pass),
        };

        client.sign_up(sign_up_req).await.unwrap();

        let login_req = LoginRequest {
            email: String::from(email),
            password: String::from(pass),
        };

        let response = client.login(login_req).await.unwrap();
        let inner= response.into_inner();

        assert_eq!(inner.user_name, user_name);
        assert_yaml_snapshot!(format!("User logged with name {}", inner.user_name));

        Ok(())
    }
}
