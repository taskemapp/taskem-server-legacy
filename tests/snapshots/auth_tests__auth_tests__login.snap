---
source: tests/auth_tests.rs
expression: "format!(\"User logged with name {}\", inner.user_name)"
---
User logged with name test-login
